feat: Governance BLS Key Registration/Deactivation/Activation with POP/discrete log as pairing check

danielntmd opened the pull request:
- Create BN254G1 library for G1 point validation on BN254 curve
- Add registerBlsKey/updateBlsKey/removeBlsKey functions to Governance
- Enforce uniqueness constraint: each BLS key can only be used by one address
- Add test suite
- Use Constants.P from ConstantsGen.sol for field modulus validation

Commit:
feat: add BLS public key storage with uniqueness validation to Governance

- Create BN254G1 library for G1 point validation on BN254 curve
- Add registerBlsKey/updateBlsKey/removeBlsKey functions to Governance
- Enforce uniqueness constraint: each BLS key can only be used by one address
- Add test suite
- Use Constants.P from ConstantsGen.sol for field modulus validation

Commit:
clean

iAmMichaelConnor reviewed (commented):
Thanks very much for turning this around so fast.

Perhaps we should take a step back and agree on the exact approach that we want for BLS key registration. I appreciate we flip-flopped and only gave a concrete opinion yesterday, so apologies for that.

The approach we've settled upon is in line with how other people in the ecosystem have built efficient BLS signatures in Solidity:

We'd ultimately like to implement Option G of this doc https://colab.research.google.com/drive/1p-BuGwcfCKEFQ2sdZTYBfCwLid2__Z_j#scrollTo=5KSbSWp3Eu7Z, but the explanation for the registration checks is in Option E (to avoid re-writing the same thing in each option).

If possible, I'd be in favour of taking some already-written code from reputable sources such as:
https://gist.github.com/kobigurk/257c1783ddf556e330f31ed57febc1d9
~https://www.zellic.io/blog/bls-signature-versatility/#modified-aggregate-public-key~
~https://github.com/Layr-Labs/eigenlayer-middleware/blob/mainnet/src/BLSApkRegistry.sol~ <-- edit: this license is a little unconventional.
https://github.com/thehubbleproject/hubble-contracts/blob/master/contracts/libs/BLS.sol

I'm not sure which of those are audited (we will re-audit regardless) and I'm not sure what their licenses are. But if their licenses are compatible with ours, and if we use their code, we get these benefits:
- They've thought about it for longer, and probably didn't have timescales as short as we do;
- They might already have been audited.
- They'll have more comprehensive testing that we can adopt.
- They might have written optimisations that we can use without having to re-invent the wheel.

What do you think?

Commit:
WIP: register skeleton

Commit:
WIP: template for aggregate verification, (use as fork)

Commit:
feat: Governance BLS Key Registration/Deactivation/Activation with POP/discrete log as pairing check

-add: registerKey takes in the keys from spec, validates on curve, field, and not infinity. Computes hash tilde as the hash to curve function with
domain separator. Computes gamma as keccak of the data, then builds the pairing check for POP/discrete log.

-add: Key Storage logic: add key, deactivate key, reactivate key, such that we keep a dense array for all validators as well as all live validators. Map logic to enforce no double key registration, one key per user and all other relevant checks. Adds indexing logic so that we keep the 0 value as a non-value and start validator indices from 1. Emits relevant events. Expose getter view functions to retrieve relevant validator data.

-add: BLS contract from Kobi and Hubble Project for BLS Fouque-Tibouchi Hash to Curve, its helpers, and constants. Add thin wrappers for pairing, creating gamma. Reorganized, cleaned and updated. Added g1 generator, point at infinity check, and keccak hashing for gamma construction.

-add: Add the original modexp file from Hubble Project with original license. Also updated.

-add: Testing suite for the Key Storage with test wrapper. Includes tests for double registration, double key registration from different users, reusing after deactivation, deactivation on unregistered keys, reactivate on unregistered keys, persistence and consistency of validator and live validator, and data integrity check, amongst others.

-*add: Testing for BLSKey registration, deletion and reactivation with real key values derived off-chain. (PR does not include typescript implementation)

Commit:
clean

Commit:
clean

Commit:
clean

Commit:
clean: proper errors

Commit:
clean
